Add ZIP file fallback.

MancunianSam opened the pull request:
There are files which aren't quite zip files which truevfs won't open.

This change tries to open the file using the truevfs classes as usual.
If there is an error doing this, it will try again with the Apache commons
zip library which is a lot less fussy about what it opens.

I've tried it against one test file I had and this is working.

I've added in a test as well which creates a zip file with one entry and
then overrides the file count byte with 0 instead of 1. This causes it
to fail on the initial read and passes when read by commons-compress.

I've also added some generic types to get rid of the warnings I keep getting about `IdentificationRequest` missing a type.

techncl reviewed (commented):
Looks good! Just a few comments. Also,

1. It's not clear in your test what you did to corrupt the Zip file (i.e. what makes the Zip file corrupt); is it possible to maybe change the variable name to have the word "corrupt" prepended/appended and/or to add a comment describing how the Zip file is corrupt?
2. There is a conditional statement `if (entry != null) {`; should we test for that
3. Should we add a test for the regular `process` i.e. if a zip file hasn't been corrupted?

MancunianSam commented on droid-container/src/main/java/uk/gov/nationalarchives/droid/container/zip/ZipIdentifierEngine.java:
```diff
@@ -43,44 +46,68 @@
                 final ZipEntry entry = zipFile.entry(entryName);
                 if (entry != null) {
                     // Get a stream for the entry and a byte reader over the stream:
                     InputStream stream = zipFile.getInputStream(entry.getName());
```
I'm going to leave it, only because there might be some weird edge case where they're not the same and I generally try to avoid touching anything that I know works.

MancunianSam commented on droid-container/src/main/java/uk/gov/nationalarchives/droid/container/zip/ZipIdentifierEngine.java:
```diff
@@ -43,44 +46,68 @@
+    }
+
+    private void processFallback(IdentificationRequest<InputStream> request, ContainerSignatureMatchCollection matches) throws IOException {
+        try (org.apache.commons.compress.archivers.zip.ZipFile zipFile = org.apache.commons.compress.archivers.zip.ZipFile.builder()
```
I avoided `var` because I thought it was like Groovy's var where everything ends up as `Object` but it's actually more like Scala's so I've changed it.

MancunianSam commented on droid-container/src/main/java/uk/gov/nationalarchives/droid/container/zip/ZipIdentifierEngine.java:
```diff
@@ -43,44 +46,68 @@
+    private void processFallback(IdentificationRequest<InputStream> request, ContainerSignatureMatchCollection matches) throws IOException {
+        try (org.apache.commons.compress.archivers.zip.ZipFile zipFile = org.apache.commons.compress.archivers.zip.ZipFile.builder()
+                .setIgnoreLocalFileHeader(true)
+                .setSeekableByteChannel(new ByteseekWindowWrapper(request.getWindowReader()))
```
There's no direct equivalent of the `preamble` and `postamble` arguments. The class documentation says that it allows you to skip the preamble which and the charset defaults to the default charset. 
So basically, those are the defaults and don't need setting.

MancunianSam commented:
1. I have added a comment as to why it's corrupt, `//14 bytes from the end is the file count. Setting it to zero causes truevfs to fail but commons-compress succeeds.`
2. Yeah, that happens if the file you're looking for isn't in the zip. I've added a test.
3. Same, I created a zip without fiddling and that passes too.
